src/content/developers/docs/data-structures/patricia-merkle-trie/index.md

```diff
@@ -0,0 +1,217 @@
+
+```
+    <64 6f> : 'verb'
+    <64 6f 67> : 'puppy'
+    <64 6f 67 65> : 'coin'
+    <68 6f 72 73 65> : 'stallion'
+```
+
+Now, we build such a trie with the following key/value pairs in the underlying DB:
+
+```
+    rootHash: [ <16>, hashA ]
+    hashA:    [ <>, <>, <>, <>, hashB, <>, <>, <>, [ <20 6f 72 73 65>, 'stallion' ], <>, <>, <>, <>, <>, <>, <>, <> ]
+    hashB:    [ <00 6f>, hashD ]
+    hashD:    [ <>, <>, <>, <>, <>, <>, hashE, <>, <>, <>, <>, <>, <>, <>, <>, <>, 'verb' ]
+    hashE:    [ <17>, [ <>, <>, <>, <>, <>, <>, [ <35>, 'coin' ], <>, <>, <>, <>, <>, <>, <>, <>, <>, 'puppy' ] ]
+```
+
+When one node is referenced inside another node, what is included is `H(rlp.encode(x))`, where `H(x) = sha3(x) if len(x) >= 32 else x` and `rlp.encode` is the [RLP](/fundamentals/rlp) encoding function.
+
+Note that when updating a trie, one needs to store the key/value pair `(sha3(x), x)` in a persistent lookup table _if_ the newly-created node has length >= 32. However, if the node is shorter than that, one does not need to store anything, since the function f(x) = x is reversible.
+
+## Tries in Ethereum {#tries-in-ethereum}
+
+All of the merkle tries in Ethereum's execution layer use a Merkle Patricia Trie.
+
+From a block header there are 3 roots from 3 of these tries.
+
+1.  stateRoot
+2.  transactionsRoot
+3.  receiptsRoot
+
+### State Trie {#state-trie}
+
+There is one global state trie, and it updates over time. In it, a `path` is always: `sha3(ethereumAddress)` and a `value` is always: `rlp(ethereumAccount)`. More specifically an ethereum `account` is a 4 item array of `[nonce,balance,storageRoot,codeHash]`. At this point it's worth noting that this `storageRoot` is the root of another patricia trie:
+
+### Storage Trie {#storage-trie}
+
+Storage trie is where _all_ contract data lives. There is a separate storage trie for each account. To calculate a 'path' in this trie first understand how solidity organizes a [variable's position](/json-rpc/API#eth_getstorageat). To get the `path` there is one extra hashing (the link does not describe this). For instance the `path` for the zeroith variable is `sha3(<0000000000000000000000000000000000000000000000000000000000000000>)` which is always `290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563`. The value at the leaf is the rlp encoding of the storage value `1234` (`0x04d2`), which is `<82 04 d2>`.
+For the mapping example, the `path` is `sha3(<6661e9d6d8b923d5bbaab1b96e1dd51ff6ea2a93520fdc9eb75d059238b8c5e9>)`
+
+### Transactions Trie {#transaction-trie}
+
+There is a separate transactions trie for every block. A `path` here is: `rlp(transactionIndex)`. `transactionIndex` is its index within the block it's mined. The ordering is mostly decided by a miner so this data is unknown until mined. After a block is mined, the transaction trie never updates.
```

Comment:
It might be interesting to add the value here since it changed recently due to [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) with legacy transactions being simply rlp-encoded and typed transactions having arbitrary encoding left-padded with the type byte.

Reply:
Which value were you referring to here @s1na?
 
Also small grammar nitpick:
```suggestion
There is a separate transactions trie for every block. A `path` here is: `rlp(transactionIndex)`. `transactionIndex` is its index within the block it's mined in. The ordering is mostly decided by a miner so this data is unknown until mined. After a block is mined, the transaction trie never updates.
```

Reply:
The transactionTrie stores (key, value) pairs where `key == rlp(transactionIndex)`, but `value` is omitted. And that would be:

```
if legacyTx:
    value = rlp(tx)
else:
    value = TxType | encode(tx)
```

Reply:
Gotcha, thanks @s1na... 
Perhaps copy as such then:

---

There is a separate transactions trie for every block. A `path` here is: `rlp(transactionIndex)`. `transactionIndex` is its index within the block it's included in. The ordering is mostly decided by a miner so this data is unknown until mined. After a block is mined, the transaction trie never updates. The value at the leaf is determined as such:

```py
if legacyTx:
  value = rlp(tx)
else:
  value = TxType | encode(tx)
```

---

_Or alternatively perhaps:_

---

There is a separate transactions trie for every block, again storing (key, value) pairs. A `path` here is: `rlp(transactionIndex)` which represents the key that corresponds to a value determined by:
```py
if legacyTx:
  value = rlp(tx)
else:
  value = TxType | encode(tx)
```
`transactionIndex` is its index within the block it's included in. The ordering is mostly decided by a miner so this data is unknown until mined. After a block is mined, the transaction trie never updates.

---

(These contains triple-backticks and have forgotten how to include/escape these in a code suggestion, sorry 😄 )

Reply:
Please also add a link to EIP-2718 for more info :)